### PR TITLE
network: extend 'enslaved' state to bridge slave

### LINKED ISSUE #5043: networkd: ethernet interfaces participating in a bridge are stuck in "configuring", failing wait-online
### Submission type

  - Bug report

### systemd version the issue has been seen with

> systemd 232
> +PAM -AUDIT -SELINUX -IMA -APPARMOR +SMACK -SYSVINIT +UTMP +LIBCRYPTSETUP +GCRYPT +GNUTLS +ACL +XZ +LZ4 +SECCOMP +BLKID +ELFUTILS +KMOD +IDN

### Used distribution

> Arch

### In case of bug report: Expected behaviour you didn't see

When an ethernet interface is linked with a bridge by a `Bridge=` directive (and no other configuration present), it should transition to "configured" as soon as it is added to that bridge.

### In case of bug report: Unexpected behaviour you saw

That interface is stuck in "configuring" state, failing wait-online.

### In case of bug report: Steps to reproduce the problem

1. Consider a system with a single Ethernet interface in a network managed by DHCPv4.

2. Place following files in `/etc/systemd/network`:
  - `/etc/systemd/network/br0.netdev`:
```
[NetDev]
Description=Bridge for libvirt
Name=br0
Kind=bridge
```
  - `/etc/systemd/network/br0.network`:
```
[Match]
Name=br0

[Network]
DHCP=ipv4

[DHCP]
UseDomains=yes
```
  - `/etc/systemd/network/ethernet.network`:
```
[Match]
Name=en*

[Network]
Bridge=br0
```

3. Restart networkd and get this output (wlp2s0 disabled, tun0 manually managed):
```
$ networkctl
IDX LINK             TYPE               OPERATIONAL SETUP
  1 lo               loopback           carrier     unmanaged
  2 enp4s0           ether              carrier     configuring
  3 wlp2s0           wlan               off         unmanaged
  4 br0              ether              routable    configured
 14 tun0             none               routable    unmanaged

5 links listed.
```

### PR DESCRIPTION
**Pull request by yuwata**
~~The change affects not only for bridge slaves, but this PR is mostly for bridge interfaces.~~

cc @ssahani.

Fixes #5043.

### CODE REVIEW

File: src/network/networkd-link.c [hunk 1]
```diff
@@ -151,7 +151,7 @@ static bool link_ipv6_enabled(Link *link) {
         if (!socket_ipv6_is_supported())
                 return false;
 
-        if (link->network->bridge || link->network->bond)
+        if (link->network->bond)
```
Comment (ssahani) [inline src/network/networkd-link.c:161]: ~~This is not right. assigning to bridge slave ipv6 addresses makes no sense.~~ @tomty89 ? 
But again kernel supports it so seems like something is not right we are doing.

Comment (yuwata) [reply src/network/networkd-link.c:161]: @ssahani and @tomty89 Yeah, It is strange to assign addresses to slave, but it works :-)

Comment (ssahani) [reply src/network/networkd-link.c:161]: @yuwata is right. LL addressing is supported by kernel. Also it anonymous (bridge does not have any addresses) bridge slave have addresses.
```
14: veth0@veth1: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default qlen 1000
    link/ether 0a:e8:65:c3:de:d8 brd ff:ff:ff:ff:ff:ff
    inet6 fe80::8e8:65ff:fec3:ded8/64 scope link tentative 
       valid_lft forever preferred_lft forever
15: veth1@veth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue master br1 state UP group default qlen 1000
    link/ether f6:f0:cd:5c:01:74 brd ff:ff:ff:ff:ff:ff
    inet6 fe80::f4f0:cdff:fe5c:174/64 scope link tentative 
       valid_lft forever preferred_lft forever
16: br1: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default qlen 1000
    link/ether f6:f0:cd:5c:01:74 brd ff:ff:ff:ff:ff:ff
    inet6 fe80::f4f0:cdff:fe5c:174/64 scope link tentative 
       valid_lft forever preferred_lft forever
```

File: test/test-network/systemd-networkd-tests.py [hunk 2]
```diff
@@ -1206,6 +1206,48 @@ def test_bridge_ignore_carrier_loss(self):
...
+
+        output = subprocess.check_output(['networkctl', 'status', 'bridge99']).rstrip().decode('utf-8')
+        self.assertRegex(output, 'State: routable \(configured\)')
+
+        output = subprocess.check_output(['networkctl', 'status', 'dummy98']).rstrip().decode('utf-8')
+        self.assertRegex(output, 'State: enslaved \(configured\)')
+
+        output = subprocess.check_output(['ip', 'rule', 'list', 'table', '100']).rstrip().decode('utf-8')
+        print(output)
+        self.assertRegex(output, 'from all to 8.8.8.8 lookup 100')
```
Comment (amishmm) [inline test/test-network/systemd-networkd-tests.py]: I am not good at python. Neither good at creating test case.

But does this verify if there are duplicate rules created?

What I believe is this just verifies that rule exists. (once or more does not matter)

Comment (yuwata) [reply test/test-network/systemd-networkd-tests.py]: Ah, yeah, right. This just checks whether the rule exists or not. But from the output, I confirm that it only shows single line, that is, not duplicated.

Comment (amishmm) [reply test/test-network/systemd-networkd-tests.py]: Ok thank you :)

**Comment by yuwata** [comment]
Hmm, the first run of CentOS CI fails
```
======================================================================
ERROR: test_dhcp_route_criticalconnection_true (__main__.NetworkdNetworkDHCPClientTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "./test/test-network/systemd-networkd-tests.py", line 1506, in test_dhcp_route_criticalconnection_true
    self.start_networkd()
  File "./test/test-network/systemd-networkd-tests.py", line 171, in start_networkd
    subprocess.check_call('systemctl start systemd-networkd', shell=True)
  File "/usr/lib64/python3.6/subprocess.py", line 291, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command 'systemctl start systemd-networkd' returned non-zero exit status 1.

======================================================================
ERROR: test_dhcp_route_metric (__main__.NetworkdNetworkDHCPClientTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "./test/test-network/systemd-networkd-tests.py", line 1494, in test_dhcp_route_metric
    self.start_networkd()
  File "./test/test-network/systemd-networkd-tests.py", line 173, in start_networkd
    subprocess.check_call('systemctl restart systemd-networkd', shell=True)
  File "/usr/lib64/python3.6/subprocess.py", line 291, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command 'systemctl restart systemd-networkd' returned non-zero exit status 1.

----------------------------------------------------------------------
Ran 74 tests in 671.674s

FAILED (errors=2, expected failures=8)
```

**Comment by yuwata** [comment]
The second run is worse...
```
======================================================================
ERROR: test_dhcp_route_criticalconnection_true (__main__.NetworkdNetworkDHCPClientTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "./test/test-network/systemd-networkd-tests.py", line 1548, in test_dhcp_route_criticalconnection_true
    self.start_networkd()
  File "./test/test-network/systemd-networkd-tests.py", line 171, in start_networkd
    subprocess.check_call('systemctl start systemd-networkd', shell=True)
  File "/usr/lib64/python3.6/subprocess.py", line 291, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command 'systemctl start systemd-networkd' returned non-zero exit status 1.

======================================================================
ERROR: test_dhcp_route_metric (__main__.NetworkdNetworkDHCPClientTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "./test/test-network/systemd-networkd-tests.py", line 1536, in test_dhcp_route_metric
    self.start_networkd()
  File "./test/test-network/systemd-networkd-tests.py", line 173, in start_networkd
    subprocess.check_call('systemctl restart systemd-networkd', shell=True)
  File "/usr/lib64/python3.6/subprocess.py", line 291, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command 'systemctl restart systemd-networkd' returned non-zero exit status 1.

======================================================================
FAIL: test_bridge_property (__main__.NetworkdNetWorkBondTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "./test/test-network/systemd-networkd-tests.py", line 1079, in test_bridge_property
    self.assertRegex(output, 'State: degraded \(configured\)')
AssertionError: Regex didn't match: 'State: degraded \\(configured\\)' not found in '● 73: bond99\n       Link File: /usr/lib/systemd/network/99-default.link\n    Network File: /run/systemd/network/bond99.network\n            Type: bond\n           State: no-carrier (configured)\n          Driver: bonding\n      HW Address: 96:36:01:87:3f:09\n         Address: fe80::9436:1ff:fe87:3f09'

----------------------------------------------------------------------
Ran 75 tests in 680.277s

FAILED (failures=1, errors=2, expected failures=8)
```
Interestingly, the last error can be also observed in #11766. See https://ci.centos.org/job/systemd-pr-build/4865/.

**Comment by yuwata** [comment]
@mrc0mmand Could you give debug logs of `networkd` when running the following tests?
- `NetworkdNetworkDHCPClientTests.test_dhcp_client_vrf`
- `NetworkdNetworkDHCPClientTests.test_dhcp_route_criticalconnection_true`

**Comment by yuwata** [comment]
@mrc0mmand Sorry, I hope I've resolved the restart issues in test-network. So, it is not necessary to provide logs.

**Comment by yuwata** [comment]
Still I cannot find the root of `NetworkdNetWorkBondTests` failure, but it disappears now. PTAL.

**Comment by yuwata** [comment]
Updated. Changes are below.
- Rebased.
- Only extend `enslaved` state to bridge slaves. See @ssahani's [comment](https://github.com/systemd/systemd/pull/11681#issuecomment-463979604).
- Update man pages.

PTAL.

File: src/network/networkd-link.c [hunk 3]
```diff
@@ -318,6 +318,19 @@ static int link_enable_ipv6(Link *link) {
         return 0;
 }
 
+static bool link_is_enslaved(Link *link) {
+        if (!link->enslaved)
+                return false;
+
+        if (link->network && link->network->bond && link->flags & IFF_SLAVE)
```
Comment (poettering) [inline src/network/networkd-link.c]: hmm, I find the mixture of && and & hard to read. maybe use FLAGS_SET() for the IFF_SLAVE check?

File: src/network/networkd-link.c [hunk 4]
```diff
@@ -318,6 +318,19 @@ static int link_enable_ipv6(Link *link) {
         return 0;
 }
 
+static bool link_is_enslaved(Link *link) {
+        if (!link->enslaved)
```
Comment (poettering) [inline src/network/networkd-link.c]: hmm, so there's now a function is_enslaved and a boolean "enslaved", and they are kinda the same but also not. I wonder if one of them should have a different name. Maybe the bool should be called "enslaved_raw" or so? (better names might be thinkable)

File: src/network/networkd-link.c [hunk 5]
```diff
@@ -651,6 +651,12 @@ static Link *link_free(Link *link) {
 
         hashmap_free(link->bond_slaves);
 
+        if (link->network) {
+                if (link->network->bond &&
+                    link_get(link->manager, link->network->bond->ifindex, &master) >= 0)
```
Comment (poettering) [inline src/network/networkd-link.c:690]: these two if expressions could be merged into one?

**Comment by poettering** [comment]
looks good. some minor comments above

File: src/network/networkd-link.c (hunk 5 above)
Comment (yuwata) [reply src/network/networkd-link.c:690]: Yeah, but a later commit adds similar expression.

File: src/network/networkd-link.c (hunk 3 above)
Comment (yuwata) [reply src/network/networkd-link.c]: Indeed. The line seems confusing :-p

**Comment by yuwata** [comment]
Updated. Changes are the following:
- Rearrange ordering in `link_is_enslaved()`
- Rename `Link::enslaved` to `Link::enslaved_raw` and add comment why it is named `_raw`.

I noticed that if a bridge slave is not managed by `networkd`, then it cannot be in enslaved state. But, I'd like to leave that for later PRs.

PTAL. Thank you.

**Pull request merged**